### PR TITLE
fix(definition-list): display as grid to enable content wrapping FE-3160

### PR DESCRIPTION
**Pull request by edleeks87**
### Proposed behaviour

Updates `DefinitionList` to `display: grid` in order to support the component being both responsive whilst also maintain alignment between  associated `Dt` and `Dd` components when the content wraps.

Adds an override to ensure the `Accordion` component has dynamic height when rendering a `DefinitionList` as a child.

Adds some examples of the component used inside an `Accordion` and in a `Grid`.
### Current behaviour

The original implementation used a single dimension to align the `Dt` and `Dd` children in columns
which caused issues with aligning them when some content wrapped. 

### Checklist

- [x] Commits follow our style guide
- [x] Screenshots are included in the PR if useful
- [x] All themes are supported if required
- [x] Unit tests added or updated if required
- [x] Cypress automation tests added or updated if required
- [x] Storybook added or updated if required
- [x] Typescript `d.ts` file added or updated if required
- [x] Carbon implementation and Design System documentation are congruent

### Testing instructions

http://localhost:9001/?path=/story/design-system-grid--subgrid
http://localhost:9001/?path=/story/design-system-tile--tile-with-definition-list-default
http://localhost:9001/?path=/story/design-system-tile--tile-with-definition-list-with-custom-width
http://localhost:9001/?path=/story/design-system-accordion--with-a-definition-list

### CODE REVIEW

**Review by jamime — COMMENTED** [review]
I've not reviewed the code, but it appears that the accordion no longer animates in the definition list example.

**Comment by edleeks87** [comment]
> I've not reviewed the code, but it appears that the accordion no longer animates in the definition list example.

So it does good spot, I'll come up with a different approach

**Review by grabkowski — APPROVED** [review]

File: src/components/accordion/accordion.style.js [hunk 1]
```diff
@@ -82,9 +82,15 @@ const StyledAccordionContentContainer = styled.div`
   box-sizing: border-box;
   overflow: hidden;
   transition: all 0.3s;
-  max-height: ${({ maxHeight }) => maxHeight}px;
-  height: ${({ maxHeight }) => maxHeight}px;
-  ${({ isExpanded }) => !isExpanded && 'visibility: hidden'};
+  ${({ maxHeight }) => css`
```
Comment (jamime) [inline src/components/accordion/accordion.style.js:85]: Please can we change this. Currently a closed accordion will output two CSS rules for `max-height` and `height`.

File: src/components/definition-list/dd.component.js [hunk 2]
```diff
@@ -5,11 +5,13 @@ import { StyledDd } from './definition-list.style';
...
 const Dd = ({
   mb = 2,
-  children
+  children,
+  align = 'right'
 }) => {
   return (
     <StyledDd
       mb={ mb }
+      align={ align }
```
Comment (jamime) [inline src/components/definition-list/dd.component.js:14]: This doesn't appear to be used anywhere. Please remove.

File: src/components/definition-list/dl.component.js [hunk 3]
```diff
@@ -4,27 +4,48 @@ import PropTypes from 'prop-types';
...
+      ddContent.push(child);
+    }
 
-      <StyledDtDiv w={ w } dtTextAlign={ dtTextAlign }>
-        {dtLabels}
-      </StyledDtDiv>
+    if (dtLabel && (index === (listChildren.length - 1) || listChildren[index + 1].type === Dt)) {
+      const key = child.props.key || `dl_row_${createGuid()}}`;
+      dlComponent.push(
+        <React.Fragment key={ key }>
```
Comment (jamime) [inline src/components/definition-list/dl.component.js:31]: `createGuid` has the same problem as using `index` in this case, as the order doesn't change, it's fine to use `index`.

https://reactjs.org/docs/lists-and-keys.html#keys

```suggestion
        <React.Fragment key={ child.props.key || index }>
```

**Review by jamime — CHANGES_REQUESTED** [review]

File: src/components/definition-list/dl.component.js (hunk 3 above)
Comment (edleeks87) [reply src/components/definition-list/dl.component.js:31]: ok thanks

File: src/components/definition-list/dd.component.js (hunk 2 above)
Comment (edleeks87) [reply src/components/definition-list/dd.component.js:14]: Consider it done

File: src/components/accordion/accordion.style.js (hunk 1 above)
Comment (edleeks87) [reply src/components/accordion/accordion.style.js:85]: Good point will update

File: src/components/accordion/accordion.spec.js [hunk 4]
```diff
@@ -140,6 +140,44 @@ describe('Accordion', () => {
...
+            </Accordion>
+          </AccordionGroup>
+        );
+      });
+
+      it('recalculates the content height', () => {
+        act(() => render({ expanded: true }));
+        wrapper.update();
+        isExpanded(wrapper);
+
```
Comment (jamime) [inline src/components/accordion/accordion.spec.js:159]: lets add an assertion here for the height before resize.

File: src/components/definition-list/dl.component.js [hunk 5]
```diff
@@ -11,20 +11,39 @@ const Dl = ({
...
+      dtLabel = child;
+    }
+    if (child.type === Dd) {
+      ddContent.push(child);
+    }
 
-      <StyledDtDiv w={ w } dtTextAlign={ dtTextAlign }>
-        {dtLabels}
-      </StyledDtDiv>
+    if (dtLabel && (index === (listChildren.length - 1) || listChildren[index + 1].type === Dt)) {
```
Comment (jamime) [inline src/components/definition-list/dl.component.js:27]:
```suggestion
    const isLastChild = index === (listChildren.length - 1);
    const nextItemIsDt = listChildren[index + 1].type === Dt;
    if (dtLabel && (nextItemIsDt || isLastChild)) {
```

**Review by jamime — CHANGES_REQUESTED** [review]

**Review by grabkowski — DISMISSED** [review]

**Review by jamime — DISMISSED** [review]

**Review by DlgSHi — APPROVED** [review]

**Review by DlgSHi — APPROVED** [review]

**Review by jamime — APPROVED** [review]

**Pull request merged**